client/allocrunner/csi_hook.go

```diff
@@ -1,46 +1,126 @@
 package allocrunner
 
 import (
+	"context"
+	"fmt"
+
 	hclog "github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/nomad/client/pluginmanager/csimanager"
 	"github.com/hashicorp/nomad/nomad/structs"
 )
 
+// RPCer is the interface needed by a csiHook to make RPC calls.
+type RPCer interface {
+	// RPC allows retrieving volume info.
+	RPC(method string, args interface{}, reply interface{}) error
+}
+
 // csiHook will wait for remote csi volumes to be attached to the host before
 // continuing.
 //
 // It is a noop for allocs that do not depend on CSI Volumes.
 type csiHook struct {
-	alloc  *structs.Allocation
-	logger hclog.Logger
+	alloc      *structs.Allocation
+	logger     hclog.Logger
+	csimanager csimanager.Manager
+	rpcClient  RPCer
+	updater    allocResourceSetter
 }
 
 func (c *csiHook) Name() string {
 	return "csi_hook"
 }
 
+func (c *csiHook) csiVolumesFromAlloc() (map[string]*structs.CSIVolume, error) {
+	vols := make(map[string]*structs.VolumeRequest)
+	tg := c.alloc.Job.LookupTaskGroup(c.alloc.TaskGroup)
+	for alias, vol := range tg.Volumes {
+		if vol.Type == structs.VolumeTypeCSI {
+			vols[alias] = vol
+		}
+	}
+
+	csiVols := make(map[string]*structs.CSIVolume, len(vols))
+	for alias, request := range vols {
+		req := &structs.CSIVolumeGetRequest{
+			ID: request.Source,
+		}
+		req.Region = c.alloc.Job.Region
+
+		var resp structs.CSIVolumeGetResponse
+		err := c.rpcClient.RPC("CSIVolume.Get", req, &resp)
```

Comment:
It looks like we need to do both `CSIVolume.Get` and `CSIVolume.Claim`. Would it maybe make more sense for us to assume that the client has a valid ID and do only a `CSIVolume.Claim`, and then change the response body for `CSIVolume.Claim` so that it includes the `structs.CSIVolume` we need?

Reply:
Probably - but if we do, I'd like us to implement something like `CSIVolume.ClaimBatch` so we can do all of the validation at once, and avoid having to unwind the claims from the client

Reply:
Good idea 👍